Combine leadership and NQT to job role in hiring staff journey

david-mears-dfe opened the pull request:
This PR adds the new job role field to the hiring staff journey.

## In this PR
- Add job role string array
- Add NQT to job role for vacancies where NQT true
- When NQT selected in job role, set NQT boolean to true
- When NQT not selected in job role, set NQT boolean to false
- Don't change search
- Don't drop NQT column yet

## Jira ticket URL
[TEVA-534](https://dfedigital.atlassian.net/browse/TEVA-534)

## Next steps
- Update jobseeker view to display job role info
- Remove leadership from code and database
- Update search to filter roles based on NQT being a job role

david-mears-dfe commented on spec/features/hiring_staff_can_publish_a_vacancy_spec.rb:
```diff
@@ -58,24 +59,22 @@
-        within_row_for(element: 'legend', text: strip_tags(I18n.t('jobs.salary_range_html'))) do
-          expect(page).to have_content(I18n.t('activerecord.errors.models.vacancy.attributes.minimum_salary.blank'))
-        end
+        mandatory_fields = %w[job_title job_role job_description salary_range working_patterns]
```
This is a refactor (DRYing) as I added job_role

david-mears-dfe commented:
It looks like I forgot to include a change to schema.rb:

```ruby
     t.uuid "publisher_user_id"
     t.datetime "expiry_time"
     t.string "supporting_documents"
+    t.string "job_role", array: true
     t.string "salary"
     t.index ["expires_on"], name: "index_vacancies_on_expires_on"
     t.index ["expiry_time"], name: "index_vacancies_on_expiry_time"
```

but I'm not adding it right now because git pulling the branch gives me a lot of merge conflicts.

cpjmcquillan commented:
> It looks like I forgot to include a change to schema.rb:
> 
> ```ruby
>      t.uuid "publisher_user_id"
>      t.datetime "expiry_time"
>      t.string "supporting_documents"
> +    t.string "job_role", array: true
>      t.string "salary"
>      t.index ["expires_on"], name: "index_vacancies_on_expires_on"
>      t.index ["expiry_time"], name: "index_vacancies_on_expiry_time"
> ```
> 
> but I'm not adding it right now because git pulling the branch gives me a lot of merge conflicts.

I've rebased the branch on master now

david-mears-dfe commented:
> > It looks like I forgot to include a change to schema.rb:
> > ```ruby
> >      t.uuid "publisher_user_id"
> >      t.datetime "expiry_time"
> >      t.string "supporting_documents"
> > +    t.string "job_role", array: true
> >      t.string "salary"
> >      t.index ["expires_on"], name: "index_vacancies_on_expires_on"
> >      t.index ["expiry_time"], name: "index_vacancies_on_expiry_time"
> > ```
> > 
> > 
> > but I'm not adding it right now because git pulling the branch gives me a lot of merge conflicts.
> 
> I've rebased the branch on master now

I deleted the comment because actually the change to schema.rb was there all along.

david-mears-dfe commented:
Ignore or delete the last merge please, it was a mistake.

cesidio commented on db/migrate/20200319165536_add_job_role_to_vacancies.rb:
```diff
@@ -0,0 +1,5 @@
+class AddJobRoleToVacancies < ActiveRecord::Migration[5.2]
+  def change
+    add_column :vacancies, :job_role, :string, array: true
```
This should be called `job_roles` everywhere, it it hasn't been deployed anywhere, I would change it before getting it merged

david-mears-dfe commented on lib/tasks/convert_nqt_to_job_role.rake:
```diff
@@ -0,0 +1,31 @@
+        )
+      end
+
+      Rails.logger.info("Conversion of salary ranges to strings has been completed for #{updated_count} vacancies")
```
This seems like the wrong message for this task

cpjmcquillan replied:
Good catch...copy and paste 😡

tatyree commented on app/controllers/hiring_staff/vacancies/job_specification_controller.rb:
```diff
@@ -50,17 +50,29 @@ def update
+
+  # Only necessary until changes to search are implemented
+  # TODO remove after migration to remove newly qualified teacher column
+  def persist_nqt_job_role_to_nqt_attribute
```
Could you abstract this please?  It's identical to the method with the same name in the copy controller except that it uses a different parameter for the job_role and adds data to a different set of parameters, both of which could be passed in.

tatyree commented on spec/lib/job_posting_spec.rb:
```diff
@@ -6,6 +6,7 @@
       '@type' => 'JobPosting',
       'title' => 'Teacher of English',
+      'jobRole' => [I18n.t('jobs.job_role_options.nqt_suitable'), I18n.t('jobs.job_role_options.sen_specialist')],
       'salary' => 'Lots of money',
```
As per comm's comment in the last show and tell, and while you're working in this part of the app, can you go ahead and change this copy here to something that doesn't worry them, please?  Maybe just some lorem.

david-mears-dfe commented on spec/controllers/hiring_staff/vacancies/job_specification_controller_spec.rb:
```diff
@@ -0,0 +1,41 @@
+        }
+      end
+
+      it 'persists the job role in the NQT field' do
```
it 'does not persist' ?

cpjmcquillan replied:
I would say it always persists the state of the job role: true or false for NQT

david-mears-dfe commented on spec/features/hiring_staff_can_edit_a_published_vacancy_spec.rb:
```diff
@@ -66,6 +76,29 @@
+
+        expect(page).to have_content(I18n.t('jobs.job_details'))
+        expect(page.find('h2', text: I18n.t('jobs.job_details'))
+          .text).to_not include(I18n.t('jobs.notification_labels.new'))
```
Maybe we should also say:

```ruby
        expect(page).not_to have_content(I18n.t('messages.jobs.new_sections.message'))
```

to verify that the new_sections message disappears?

david-mears-dfe commented on app/controllers/hiring_staff/vacancies/copy_controller.rb:
```diff
@@ -23,12 +22,26 @@ def create
   private
 
   def copy_form_params
+    persist_nqt_job_role_to_nqt_attribute
```
As there is this complexity here, it's probably worth adding something to `spec/features/hiring_staff_can_copy_a_vacancy_spec.rb` to ensure that copying a vacancy where the job_role includes nqt results in a new vacancy with newly_qualified_teacher == true.

david-mears-dfe replied:
Separately, while you are touching `spec/features/hiring_staff_can_copy_a_vacancy_spec.rb`, I suspect we should update the lists of attributes expected to have been copied.

cpjmcquillan replied:
> As there is this complexity here, it's probably worth adding something to `spec/features/hiring_staff_can_copy_a_vacancy_spec.rb` to ensure that copying a vacancy where the job_role includes nqt results in a new vacancy with newly_qualified_teacher == true.

Given this method is abstracted, and its behaviour is tested elsewhere, and the NQT field itself will be deleted soon, I would err against adding this to the feature spec

cpjmcquillan replied:
> Separately, while you are touching `spec/features/hiring_staff_can_copy_a_vacancy_spec.rb`, I suspect we should update the lists of attributes expected to have been copied.

I don't think this is necessary
